### PR TITLE
cli: structure CLI report handling

### PR DESCRIPTION
**Pull request by haxscramper**
Structure data pass for the command-line report generation: `--help`,
`--fullhelp`, `--version`. Remove raw usage of `writeln` from
`commands.nim`.

### CODE REVIEW

**Comment by haxscramper** [comment]
bors try

File: compiler/cli_reporter.nim [hunk 1]
```diff
@@ -2664,9 +2665,52 @@ proc reportShort*(conf: ConfigRef, r: ParserReport): string =
   reportBody(conf, r) & suffixShort(conf, r)
 
 
+proc genFeatureDesc[T: enum](t: typedesc[T]): string {.compileTime.} =
+  result = ""
+  for f in T:
+    if result.len > 0: result.add "|"
+    result.add $f
+
+const
```
Comment (saem) [inline compiler/cli_reporter.nim:2674]: Admittedly, this feels weird in here, but I got nothing actionable

**Comment by saem** [comment]
Aside: the warnings are being repeated a bunch, I think that's a regression. 🤔

**Comment by haxscramper** [comment]
Which warnings specifically you are talking about?

File: compiler/cli_reporter.nim (hunk 1 above)
Comment (haxscramper) [reply compiler/cli_reporter.nim:2674]: In the future we would get rid of this compile-time string hardcoding and build CLI `--help` interface from some DSL/object, so for now it is sufficient, in the future it won't/ be needed.

**Comment by saem** [comment]
I scrolled down the Files Changed tab, there are 11 copies of a warning in msgs.

**Comment by haxscramper** [comment]
Those are separate reports from different matrix configurations, no?

image

Terminal import need to be removed, but otherwise it is perfectly fine.

**Comment by saem** [comment]
Ah, hard to notice on the phone, thanks

File: compiler/msgs.nim [hunk 2]
```diff
@@ -346,33 +346,6 @@ proc msgWrite*(conf: ConfigRef; s: string, flags: MsgFlags = {}) =
...
-
-macro callStyledWriteLineStderr(args: varargs[typed]): untyped =
-  result = newCall(bindSym"styledWriteLine")
-  result.add(bindSym"stderr")
-  for arg in children(args[0][1]):
-    result.add(arg)
-  when false:
-    # not needed because styledWriteLine already ends with resetAttributes
-    result = newStmtList(result, newCall(bindSym"resetAttributes", bindSym"stderr"))
-
```
Comment (alaviss) [inline compiler/msgs.nim:375]: It seems to me that this change has nothing to do with this PR and is not explained in the commit either. Should be moved to another PR or explained in the commit.

**Review by alaviss — CHANGES_REQUESTED** [review]
Looks good to me, sans a minor detail

File: compiler/msgs.nim (hunk 2 above)
Comment (haxscramper) [reply compiler/msgs.nim:375]: Updated commit message

File: compiler/cli_reporter.nim [hunk 3]
```diff
@@ -3398,12 +3442,17 @@ proc reportHook*(conf: ConfigRef, r: Report): TErrorHandling =
...
-  if conf.isEnabled(r) and r.category == repDebug and tryhack:
-    # Force write of the report messages using regular stdout if tryhack is
-    # enabled
+  if (conf.isEnabled(r) and r.category == repDebug and tryhack) or
+     # Force write of the report messages using regular stdout if tryhack is
+     # enabled
+     r.kind in rintCliKinds
+     # or if we are writing command-line help/usage information - it must
+     # always be printed
+    :
```
Comment (alaviss) [inline compiler/cli_reporter.nim]: Not really a big fan of this dangling `:`, maybe you can group the entire expression in `()` to increase clarity? Like this

```nim
if (
  longExpr1 or
  anotherLongExpr
):
  doStuff
```

File: compiler/commands.nim [hunk 4]
```diff
@@ -41,85 +41,69 @@ type
...
+      usedTinyC,
+      useLinenoise,
+      usedFFI,
+      usedBoehm,
+      usedMarkAndSweep,
+      usedGoGC,
+      usedNoGC
+    ],
+    cpu: conf.target.hostCPU,
+    os: conf.target.hostOS)
```
Comment (alaviss) [inline compiler/commands.nim]: Styling nit

```suggestion
    os: conf.target.hostOS
  )
```

File: compiler/commands.nim [hunk 5]
```diff
@@ -41,85 +41,69 @@ type
...
+
 
 proc helpOnError(conf: ConfigRef; pass: TCmdLinePass) =
   if pass == passCmd1:
-    conf.writeln(getCommandLineDesc(conf), {msgStdout})
+    conf.localReport(
```
Comment (alaviss) [inline compiler/commands.nim]: Maybe employ the block form here instead?

```suggestion
    conf.localReport:
```

File: compiler/commands.nim [hunk 6]
```diff
@@ -41,85 +41,69 @@ type
...
 
 proc writeAdvancedUsage(conf: ConfigRef; pass: TCmdLinePass) =
   if pass == passCmd1:
-    conf.writeln((
-      HelpMessage % [
-        VersionAsString,
-        platform.OS[conf.target.hostOS].name,
-        CPU[conf.target.hostCPU].name
-    ]) & AdvancedUsage, {msgStdout})
+    conf.localReport(InternalReport(
```
Comment (alaviss) [inline compiler/commands.nim]: Block form would be easier on the eyes here

```suggestion
    conf.localReport:
      InternalReport(
```

File: compiler/commands.nim [hunk 7]
```diff
@@ -41,85 +41,69 @@ type
...
     msgQuit(0)
 
 proc writeFullhelp(conf: ConfigRef; pass: TCmdLinePass) =
   if pass == passCmd1:
-    conf.writeln((HelpMessage % [
-      VersionAsString,
-      platform.OS[conf.target.hostOS].name,
-      CPU[conf.target.hostCPU].name
-    ]) & Usage & AdvancedUsage, {msgStdout})
+    conf.localReport(InternalReport(
```
Comment (alaviss) [inline compiler/commands.nim]:
```suggestion
    conf.localReport:
      InternalReport(
```

**Review by alaviss — APPROVED** [review]
LGTM

There are a couple of styling nits in case you want to deal with them.

**Comment by haxscramper** [comment]
Yes, I will fix them now

**Comment by alaviss** [comment]
I guess bors doesn't understand edits

bors r+
